crates/accelerate/src/sabre_swap/mod.rs

```diff
@@ -390,46 +496,119 @@ fn update_route(
+            )
+        })
+        .collect();
+
+    let swaps = token_swapper(
+        coupling,
+        mapping,
+        Some(SWAP_EPILOGUE_TRIALS),
+        Some(seed),
+        None,
+    );
+
+    // Convert physical swaps to virtual swaps
+    swaps
+        .into_iter()
+        .map(|(l, r)| {
+            let ret = [
+                from_layout.phys_to_logic[l.index()],
+                from_layout.phys_to_logic[r.index()],
+            ];
+            from_layout.swap_physical(l.index(), r.index());
+            ret
+        })
+        .collect()
+}
+
 /// Search forwards in the `dag` from all the nodes in `to_visit`, adding them to the `gate_order`
 /// or the current `front_layer` as appropriate, and continue inspecting gates until there is
 /// nothing further with no required predecessors.
 ///
 /// The nodes in `to_visit` should all already have no further required predecessors.
-fn route_reachable_nodes(
+fn route_reachable_nodes<F>(
+    seed: u64,
     to_visit: &[NodeIndex],
     dag: &SabreDAG,
     layout: &NLayout,
     coupling: &DiGraph<(), ()>,
     gate_order: &mut Vec<usize>,
     front_layer: &mut FrontLayer,
     required_predecessors: &mut [u32],
-) {
+    node_block_results: &mut HashMap<usize, Vec<BlockResult>>,
+    route_block_dag: &F,
+) where
+    F: Fn(&SabreDAG, NLayout) -> (SabreResult, NLayout),
+{
     let mut to_visit = to_visit.to_vec();
     let mut i = 0;
     // Iterate through `to_visit`, except we often push new nodes onto the end of it.
     while i < to_visit.len() {
         let node = to_visit[i];
         i += 1;
         let (py_node, qubits) = &dag.dag[node];
-        match qubits[..] {
-            [a, b]
-                if !coupling.contains_edge(
-                    NodeIndex::new(layout.logic_to_phys[a]),
-                    NodeIndex::new(layout.logic_to_phys[b]),
-                ) =>
-            {
-                front_layer.insert(node, [a, b]);
+
+        match dag.node_blocks.get(py_node) {
+            Some(blocks) => {
+                // Control flow op. Route all blocks for current layout.
+                let mut block_results: Vec<BlockResult> = Vec::with_capacity(blocks.len());
+                for inner_dag in blocks {
+                    let (inner_dag_routed, inner_final_layout) =
+                        route_block_dag(inner_dag, layout.copy());
+
+                    // For now, we always append a swap circuit that gets the inner block
+                    // back to the parent's layout.
+                    let swap_epilogue =
+                        gen_swap_epilogue(coupling, inner_final_layout, layout, seed);
+                    let block_result = BlockResult {
+                        result: inner_dag_routed,
+                        swap_epilogue,
+                    };
+                    block_results.push(block_result);
+                }
+                node_block_results.insert_unique_unchecked(*py_node, block_results);
             }
```

Comment:
Yes absolutely, good point, there's certainly an implicit decision here. I thought about this and figured that it shouldn't matter for this implementation, since placement of a control flow op doesn't permute the layout. If it did, I imagine it'd be best to go for ALAP, i.e. place all gate sequences that do not have a data dependency on the control flow op in question, since the lookahead would have been driving towards making these gates as close as possible, and only then place the control flow op.

I've got some notes set aside from when I was planning the work in this PR that might come in handy when we revisit this to look for optimizations.